File: leetcode/121_easy_best-time-to-buy-and-sell-stock.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
//Accepted. average space & time
#include <vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int max_profit_so_far = 0, min_price_so_far = __INT_MAX__;
        for(int i = 0; i < prices.size(); ++i){
          if (prices[i] < min_price_so_far){
            min_price_so_far = prices[i];
          }
          int current_profit = (prices[i] - min_price_so_far);
          if(current_profit > max_profit_so_far){
            max_profit_so_far = current_profit;
          }
        }
        return max_profit_so_far;
    }
};

File: leetcode/121_easy_best-time-to-buy-and-sell-stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
from types import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        bestProfit = 0
        lowestPrice = prices[0]
        for i in range(len(prices)):
            lowestPrice = min(lowestPrice, prices[i])
            bestProfit = max(bestProfit, prices[i] - lowestPrice)
        return bestProfit
